rfd/0163-vnet.md

```diff
@@ -0,0 +1,326 @@
+    1. If the app does exist, a free IP will be assigned to it and it will be
+       returned in an authoritative answer.
+1. If the name matches `*.<custom-dns-zone>`:
+    1. VNet will determine which cluster(s) the matching custom DNS zone is in from
+       its local state.
+    1. VNet will query the cluster(s) apps to see if any have a `spec.vnet_addr`
+       matching the full name. This will use a predicate expression to do the
+       filtering server-side and avoid listing all apps.
+    1. If a match is found a free IP will be assigned to that app and it will be
+       returned in an authoritative answer.
+    1. If no match is found and the zone has no `upstream_nameservers` configured,
+       no answer will be sent.
+    1. The full DNS query will be forwarded to one of the `upstream_nameservers`,
+       selected randomly from the list, and any upstream response will be
+       returned to the client verbatim.
+       We may implement caching of upstream responses, but hopefully the
+       downstream resolver will do its own caching.
+1. If the name did not match any `<proxy-address>.internal` or custom DNS zone,
+   no response will be sent.
+
+TODO: sequence diagram
+
+### IP range
+
+By default IPs will be assigned from the carrier-grade NAT range 100.64.0.0/10
+Since VNet IPs will only be used locally on end user machines this range is
+quite unlikely to conflict with any valid public or private IP addresses, unless
+another VPN client is being used that happens to use the same range.
+For example, tailscale also uses this range.
+This range can be overriden with a cluster-wide configuration.
+
+```yaml
+version: v1
+kind: vnet
+metadata:
+  name: vnet
+spec:
+  cidr_range: 100.64.0.0/10
+```
+
+### Host configuration
+
+Vnet will need to create a TUN interface, assign an IP to it, and modify the
+routing table.
+
+To create the TUN interface we will import `golang.zx2c4.com/wireguard/tun`
+which provides cross-platform methods for creating TUN interfaces and reading
+from + writing them.
+It supports MacOS, Windows, and Linux.
+
+On MacOS, to assign an IP address to the interface, we will use `os/exec` to run
+the `ifconfig` command:
+
+```shell
+ifconfig utun4 100.64.0.1 100.64.0.1 up
+```
+
+On MacOS, to assign an IP route to use the TUN interface, we will use `os/exec`
+to run the `route` command:
+
+```shell
+route add -net 100.64.0.0/10 -interface utun4
+```
+
+All system configuration must run as a privileged user/root.
+In the first version we will use an Applescript wrapper to call a child process
+as root.
+The child admin process will still be the `tsh` binary, just called with
+arguments to do the host setup.
+See the UX section for more details on the admin process.
+
+The first version of VNet will only support MacOS so this is written with that
+in mind, but we can implement this in a similar way for Windows and Linux.
+
+### Security
+
+### UX
```

Comment:
If VNet detects that certs have expired, can it prompt the user to relogin via Connect?

Reply:
Yeah, I just didn't get around to writing it down yet. We'll likely need to add some kind of middleware, similar to [`alpnproxy.LocalProxyMiddleware`](https://github.com/gravitational/teleport/blob/70ce5dc3dbd43bb96195dea2af2b163222f0c656/lib/srv/alpnproxy/local_proxy.go#L97-L105). Here's [how `lib/teleterm` utilizes it for app gateways](https://github.com/gravitational/teleport/blob/70ce5dc3dbd43bb96195dea2af2b163222f0c656/lib/teleterm/gateway/app_middleware.go) for example.